### PR TITLE
DYT-3064: Fix cross-window entity count inflation in BatchResult

### PR DESCRIPTION
**Pull request by msrdic**
## Summary
- Fixed entity count inflation in `BatchResult` when processing multi-window batches via `submit_batch`
- The root cause: `results["entities"]` was incremented by the raw entity list length for each window, counting entities that span multiple windows multiple times (since `upsert_entities_batch` correctly deduplicates at the DB level, but the counter did not)
- Fix: track a `_seen_entity_keys: set[tuple[str, str]]` across windows and only increment the counter for entities not yet seen — aligning `BatchResult.entities` with the actual number of unique entities persisted

## Test plan
- [x] Unit tests added in `tests/unit/engines/test_vectorcypher_bugs.py` covering: single-window (baseline), multi-window with distinct entities, multi-window with fully overlapping entities, and multi-window with partially overlapping entities
- [x] Integration test added in `tests/integration/test_windowed_processing_submit_batch.py` verifying the fix end-to-end via a real `submit_batch` call
- [x] All unit tests pass (`make test`)
- [x] Formatting passes (`make format`)

### CODE REVIEW

**Comment by msrdic** [comment]
## Audit Summary

| Audit | Status |
|-------|--------|
| Self-audit (`/audit --auto`) | clean (fixed in 2 cycles) |
| Specialist: `/review:test --auto` | clean (fixed in 2 cycles) |
| Bandit (Python security) | clean (fixed in 2 cycles) |
| Cross-audit (`/codex-audit`) | ran |

**Cycles:** 2/5
**Result:** clean (fixed in 2 cycles)

---

## Fix Cycle 1

### Files changed
```
src/khora/engines/vectorcypher/engine.py     | 4 ++--
 tests/unit/engines/test_vectorcypher_bugs.py | 8 ++++----
 2 files changed, 6 insertions(+), 6 deletions(-)
```

### Diff
```diff
diff --git a/src/khora/engines/vectorcypher/engine.py b/src/khora/engines/vectorcypher/engine.py
index d1f4f89..5420336 100644
--- a/src/khora/engines/vectorcypher/engine.py
+++ b/src/khora/engines/vectorcypher/engine.py
@@ -2144,7 +2144,7 @@ class VectorCypherEngine:
         # Track unique entity keys across windows to avoid double-counting entities
         # that appear in multiple windows (upsert_entities_batch ensures a single DB
         # row, so BatchResult.entities must reflect unique persisted cardinality).
-        _seen_entity_keys: set[str] = set()
+        _seen_entity_keys: set[tuple[str, str]] = set()
 
         for window_states in windows:
             # ── Stage 2: Batch-embed ALL chunk texts ────────────────────────
@@ -2456,7 +2456,7 @@ class VectorCypherEngine:
 
             new_entity_count = 0
             for _e in all_entities:
-                _key = f"{_e.name}:{_e.entity_type}"
+                _key = (_e.name, _e.entity_type)
                 if _key not in _seen_entity_keys:
                     _seen_entity_keys.add(_key)
                     new_entity_count += 1
```

### Validation

| Metric | Count |
|--------|-------|
| Fixed (verified) | 1 |
| Fixed (unverified) | 0 |
| Accepted | 0 |
| Skipped | 0 |
| Missing evidence | 0 |
| **Total expected** | **1** |

**Per-finding status:**

- M1: FIXED (verified)

---

## Audit Outcome: VERIFIED FIXED

All HIGH/MEDIUM findings from earlier cycles were verified fixed by a clean audit in cycle 2.

# Audit Findings — DYT-3064 (Cycle 2 — Verification)

Branch: `ms/DYT-3064-khora-cross-window-entity-count`
Date: 2026-04-25
Mode: Verification audit (cycle 2) — verifying prior cycle's findings only

## Verified Resolved

### M1 — String key collision in cross-window deduplication set ✅ RESOLVED
`engine.py:2459` now uses `_key = (_e.name, _e.entity_type)` (tuple). The string
concatenation bug is gone. The unit test at `test_seen_entity_keys_dedup_logic` also
uses the tuple form. **No longer a finding.**

---

## HIGH Findings

None.

---

## MEDIUM Findings

None.

---

## LOW Findings

### L1 — Duplicate `_FakeEntity` inner class across test methods (STILL PRESENT)
**File:** `tests/unit/engines/test_vectorcypher_bugs.py:385` and `:416`

`_FakeEntity` is defined identically inside both `test_seen_entity_keys_dedup_logic`
(lines 385–388) and `test_no_cross_window_inflation_when_all_distinct` (lines 416–419).
Not fixed in cycle 1. Extract to a shared class-level or module-level stub to avoid
divergence.

### L2 — No test case for the colon-collision edge case (STILL PRESENT)
**File:** `tests/unit/engines/test_vectorcypher_bugs.py`

No test exercises `name='foo:bar', entity_type='BAZ'` vs `name='foo', entity_type='bar:BAZ'`.
The underlying M1 bug is fixed (tuple key), so no regression can occur, but a regression
test would lock in the correct behaviour. Not added in cycle 1.

### L3 — Module docstring not updated for scenario 9 (STILL PRESENT)
**File:** `tests/integration/test_windowed_processing_submit_batch.py:3`

Line 3 still reads `Eight scenarios:` and lists only scenarios 1–8. Scenario 9
(`test_cross_window_entity_count_not_inflated`, line 236) is present in the file but
absent from the docstring. Not fixed in cycle 1.

### L4 — Brittle source-inspection test (STILL PRESENT, not blocking)
**File:** `tests/unit/engines/test_vectorcypher_bugs.py:358-372`

`test_inflated_accumulation_pattern_removed` uses `inspect.getsource()` to assert a
specific string literal is absent. If the engine is refactored the test may give a
false result. Pre-existing pattern, noted for awareness. Not fixed in cycle 1.

---

## Summary

| Severity | Count |
|----------|-------|
| HIGH     | 0     |
| MEDIUM   | 0     |
| LOW      | 4     |
| **Total**| **4** |

**Assessment: All MEDIUM findings resolved; only LOW findings remain.**
M1 (string key collision) was correctly fixed with a tuple key. L1–L4 are unchanged
from cycle 1. No blocking issues.

**Comment by msrdic** [comment]
**Audits** — Fixed 1 finding, 4 LOWs noted (2 cycles)

Fixed:
- 🟡 M1: String key collision in cross-window deduplication set — fixed

Noted (advisory):
- 🟢 L1: Duplicate `_FakeEntity` inner class across test methods — noted
- 🟢 L2: No test case for the colon-collision edge case — noted
- 🟢 L3: Module docstring not updated for scenario 9 — noted
- 🟢 L4: Brittle source-inspection test — noted
